src/Uno.UI/UI/Xaml/Controls/PasswordBox/PasswordBox.cs

```diff
@@ -150,21 +156,99 @@ public bool IsPasswordRevealButtonEnabled
 
 		private void OnIsPasswordRevealButtonEnabledChanged(DependencyPropertyChangedEventArgs e)
 		{
-			_isButtonEnabled = IsPasswordRevealButtonEnabled;
+			CheckRevealModeForScope();
+			OnIsPasswordRevealButtonEnabledChangedPartial(e);
+		}
+
+		partial void OnIsPasswordRevealButtonEnabledChangedPartial(DependencyPropertyChangedEventArgs e);
+		#endregion
+
+		#region PasswordRevealMode DependencyProperty
+		public PasswordRevealMode PasswordRevealMode
+		{
+			get => (PasswordRevealMode)this.GetValue(PasswordRevealModeProperty);
+			set => this.SetValue(PasswordRevealModeProperty, value);
+		}
+
+		public static global::Windows.UI.Xaml.DependencyProperty PasswordRevealModeProperty { get; } =
+			DependencyProperty.Register(
+				nameof(PasswordRevealMode),
+				typeof(bool),
+				typeof(PasswordBox),
+				new FrameworkPropertyMetadata(
+					defaultValue: PasswordRevealMode.Peek,
+					propertyChangedCallback: (s, e) => ((PasswordBox)s)?.OnPasswordRevealModeChanged(e)
+				)
+			);
+
+		private void OnPasswordRevealModeChanged(DependencyPropertyChangedEventArgs e)
+		{
+			CheckRevealModeForScope();
+		}
 
-			if (IsPasswordRevealButtonEnabled)
+		private void CheckRevealModeForScope()
+		{
+			// Only use IsPasswordRevealButtonEnabled if it is set and PasswordRevealMode is not
+			if (UseIsPasswordEnabledProperty)
 			{
-				VisualStateManager.GoToState(this, TextBoxConstants.ButtonVisibleStateName, true);
-			}
-			else
+				SetPasswordScope(true);
+			} else
 			{
-				VisualStateManager.GoToState(this, TextBoxConstants.ButtonCollapsedStateName, true);
+				switch (PasswordRevealMode)
+				{
+					case PasswordRevealMode.Visible:
+						SetPasswordScope(false);
+						break;
+					case PasswordRevealMode.Hidden:
+					case PasswordRevealMode.Peek:
+					default:
+						SetPasswordScope(true);
+						break;
+				}
 			}
+		}
+		#endregion
 
-			OnIsPasswordRevealButtonEnabledChangedPartial(e);
+		internal override void UpdateFocusState(FocusState focusState)
+		{
+			var oldValue = FocusState;
+			base.UpdateFocusState(focusState);
+			OnFocusStateChanged(oldValue, focusState);
 		}
 
-		partial void OnIsPasswordRevealButtonEnabledChangedPartial(DependencyPropertyChangedEventArgs e);
-		#endregion
+		private void OnFocusStateChanged(FocusState oldValue, FocusState newValue)
+		{
+			if (oldValue == newValue) return;
```

Comment:
Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5940293)